File: src/kibana_provider.py
from elasticsearch import Elasticsearch
from datetime import datetime, timedelta
import os
import json
from typing import Dict, List, Optional

class KibanaKeptnProvider:
    def __init__(self, es_host: str, es_port: int = 9200, 
                 username: Optional[str] = None, password: Optional[str] = None):
        self.es_client = Elasticsearch(
            [{'host': es_host, 'port': es_port}],
            http_auth=(username, password) if username and password else None
        )
        
    def query_kibana_metrics(self, index_pattern: str, metric_name: str, 
                           time_range_minutes: int = 60) -> Dict:
        """
        Query metrics from Kibana/Elasticsearch
        
        Args:
            index_pattern: The Elasticsearch index pattern to query
            metric_name: Name of the metric to retrieve
            time_range_minutes: Time range in minutes to look back
            
        Returns:
            Dict containing the query results
        """
        now = datetime.utcnow()
        time_from = now - timedelta(minutes=time_range_minutes)
        
        query = {
            "query": {
                "bool": {
                    "must": [
                        {"match": {"metric_name": metric_name}},
                        {
                            "range": {
                                "@timestamp": {
                                    "gte": time_from.isoformat(),
                                    "lte": now.isoformat()
                                }
                            }
                        }
                    ]
                }
            }
        }
        
        try:
            response = self.es_client.search(
                index=index_pattern,
                body=query
            )
            return response
        except Exception as e:
            print(f"Error querying Elasticsearch: {str(e)}")
            return {}

    def format_keptn_metrics(self, es_response: Dict) -> List[Dict]:
        """
        Format Elasticsearch response into Keptn metrics format
        
        Args:
            es_response: Response from Elasticsearch query
            
        Returns:
            List of formatted metrics for Keptn
        """
        keptn_metrics = []
        
        if 'hits' not in es_response or 'hits' not in es_response['hits']:
            return keptn_metrics
            
        for hit in es_response['hits']['hits']:
            source = hit['_source']
            
            metric = {
                'name': source.get('metric_name'),
                'value': source.get('value'),
                'timestamp': source.get('@timestamp'),
                'labels': {
                    'source': 'kibana',
                    'index': hit['_index']
                }
            }
            keptn_metrics.append(metric)
            
        return keptn_metrics

    def send_to_keptn(self, metrics: List[Dict], keptn_endpoint: str) -> bool:
        """
        Send formatted metrics to Keptn
        
        Args:
            metrics: List of formatted metrics
            keptn_endpoint: Keptn API endpoint
            
        Returns:
            bool indicating success/failure
        """
        try:
            # Here you would implement the actual API call to Keptn
            # This is a placeholder for the actual implementation
            print(f"Sending metrics to Keptn: {json.dumps(metrics, indent=2)}")
            return True
        except Exception as e:
            print(f"Error sending metrics to Keptn: {str(e)}")
            return False

File: pkg/provider/kibana_provider.go
package provider

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"time"

	"github.com/elastic/go-elasticsearch/v8"
	"github.com/sumukshashidhar/integrate-kibana-and-keptn/pkg/models"
)

type KibanaProvider struct {
	esClient *elasticsearch.Client
}

func NewKibanaProvider(esHost string, esPort int, username, password string) (*KibanaProvider, error) {
	cfg := elasticsearch.Config{
		Addresses: []string{fmt.Sprintf("http://%s:%d", esHost, esPort)},
		Username:  username,
		Password:  password,
	}

	client, err := elasticsearch.NewClient(cfg)
	if err != nil {
		return nil, fmt.Errorf("error creating elasticsearch client: %w", err)
	}

	return &KibanaProvider{
		esClient: client,
	}, nil
}

func (p *KibanaProvider) QueryKibanaMetrics(ctx context.Context, indexPattern, metricName string, timeRangeMinutes int) (*models.ElasticsearchResponse, error) {
	now := time.Now().UTC()
	timeFrom := now.Add(-time.Duration(timeRangeMinutes) * time.Minute)

	query := map[string]interface{}{
		"query": map[string]interface{}{
			"bool": map[string]interface{}{
				"must": []map[string]interface{}{
					{
						"match": map[string]interface{}{
							"metric_name": metricName,
						},
					},
					{
						"range": map[string]interface{}{
							"@timestamp": map[string]interface{}{
								"gte": timeFrom.Format(time.RFC3339),
								"lte": now.Format(time.RFC3339),
							},
						},
					},
				},
			},
		},
	}

	var buf bytes.Buffer
	if err := json.NewEncoder(&buf).Encode(query); err != nil {
		return nil, fmt.Errorf("error encoding query: %w", err)
	}

	res, err := p.esClient.Search(
		p.esClient.Search.WithContext(ctx),
		p.esClient.Search.WithIndex(indexPattern),
		p.esClient.Search.WithBody(&buf),
	)
	if err != nil {
		return nil, fmt.Errorf("error executing search: %w", err)
	}
	defer res.Body.Close()

	if res.IsError() {
		return nil, fmt.Errorf("elasticsearch error: %s", res.String())
	}

	var response models.ElasticsearchResponse
	if err := json.NewDecoder(res.Body).Decode(&response); err != nil {
		return nil, fmt.Errorf("error parsing response: %w", err)
	}

	return &response, nil
}

func (p *KibanaProvider) FormatKeptnMetrics(esResponse *models.ElasticsearchResponse) []models.KeptnMetric {
	var keptnMetrics []models.KeptnMetric

	for _, hit := range esResponse.Hits.Hits {
		metric := models.KeptnMetric{
			Name:      hit.Source.MetricName,
			Value:     hit.Source.Value,
			Timestamp: hit.Source.Timestamp,
			Labels: map[string]string{
				"source": "kibana",
				"index":  hit.Index,
			},
		}
		keptnMetrics = append(keptnMetrics, metric)
	}

	return keptnMetrics
}

func (p *KibanaProvider) SendToKeptn(ctx context.Context, metrics []models.KeptnMetric, keptnEndpoint string) error {
	payload, err := json.Marshal(metrics)
	if err != nil {
		return fmt.Errorf("error marshaling metrics: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, "POST", keptnEndpoint, bytes.NewBuffer(payload))
	if err != nil {
		return fmt.Errorf("error creating request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return fmt.Errorf("error sending metrics to Keptn: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("unexpected status code from Keptn: %d", resp.StatusCode)
	}

	return nil
}